test/test_buffer_protocol.py

```diff
@@ -0,0 +1,163 @@
+            _run_test_and_check_grad(True, dtype, [0, 1, 2, 3])
+            _run_test_and_check_grad(True, dtype, [0] * 100)
+        _run_test_and_check_grad(False, dtype, [1])
+        _run_test_and_check_grad(False, dtype, [0, 1, 2, 3])
+        _run_test_and_check_grad(False, dtype, [0] * 100)
+
+    @onlyCPU
+    @dtypes(*all_types())
+    def test_with_offset(self, device, dtype):
+        input = self.INPUT
+        # Offset should be valid whenever there is, at least,
+        # one remaining element
+        for i in range(len(input)):
+            self._run_test(dtype, input, first=i)
+
+    @onlyCPU
+    @dtypes(*all_types())
+    def test_with_count(self, device, dtype):
+        input = self.INPUT
+        # Count should be valid for any valid in the interval
+        # [-1, len(input)], except for 0
+        for i in range(-1, len(input) + 1):
+            if i != 0:
+                self._run_test(dtype, input, count=i)
+
+    @onlyCPU
+    @dtypes(*all_types())
+    def test_with_count_and_offset(self, device, dtype):
+        input = self.INPUT
+        # Explicit default count [-1, 1, 2, ..., len]
+        for i in range(-1, len(input) + 1):
+            if i != 0:
+                self._run_test(dtype, input, count=i)
+        # Explicit default offset [0, 1, ..., len - 1]
+        for i in range(len(input)):
+            self._run_test(dtype, input, first=i)
+        # All possible combinations of count and dtype aligned
+        # offset for 'input'
+        # count:[1, 2, ..., len - 1] x first:[0, 1, ..., len - count]
+        for i in range(1, len(input)):
+            for j in range(len(input) - i + 1):
+                self._run_test(dtype, input, count=i, first=j)
+
+    @onlyCPU
+    @dtypes(*all_types())
+    def test_invalid_positional_args(self, device, dtype):
+        input = self.INPUT
+        bytes = get_dtype_size(dtype)
+        # Empty array
+        with self.assertRaises(ValueError):
+            self._run_test(dtype, [])
+        # Count equals 0
+        with self.assertRaises(ValueError):
+            self._run_test(dtype, input, count=0)
+        # Offset negative and bigger than total length
+        with self.assertRaises(ValueError):
+            self._run_test(dtype, input, first=-1)
+        with self.assertRaises(ValueError):
+            self._run_test(dtype, input, first=len(input))
+        # Non-multiple offset with all elements
+        if bytes > 1:
+            with self.assertRaises(ValueError):
+                self._run_test(dtype, input, offset=bytes - 1)
+        # Count too big for each good first element
+        for first in range(len(input)):
+            with self.assertRaises(ValueError):
+                count = len(input) - first + 1
+                self._run_test(dtype, input, count=count, first=first)
+
+    @onlyCPU
+    @dtypes(*all_types())
+    def test_invalid_device(self, device, dtype):
+        input = self.INPUT
+        with self.assertRaises(RuntimeError):
```

Comment:
What is the error you catch here? That there are no cuda devices available?

Reply:
Calling `TensorMaker::for_blob` saying the data pointer lives on CUDA, ends up calling `cudaPointerGetAttributes`. If we pass a pointer that was allocated on CPU, as far as I understand it, it will error out, which will be caught by `AT_CUDA_CHECK`.

Reply:
> If we pass a pointer that was allocated on CPU, as far as I understand it, it will error out, which will be caught by `AT_CUDA_CHECK`.

Apparently, this is not true. It depends on the implementation of `cudaPointerGetAttributes` (not sure what it does when, say, it returns `cudaMemoryTypeUnregistered` as its `type`), and the context that created the pointer. Maybe it's not a good idea to have this test, then!?

Reply:
I feel like this test might segfault yes.
I think in this case, you just want to make the doc scary enough that users don't do it :D